common/web/gesture-recognizer/src/inputSequence.ts

```diff
@@ -0,0 +1,51 @@
+/// <reference path="inputSample.ts" />
+namespace com.keyman.osk {
+  type JSONInputSequence = Omit<Omit<InputSequence, '_currentTarget'>, '_identifier'>;
+
+  export class InputSequence {
+    public readonly isFromTouch: boolean;
+    public readonly identifier: number;
+
+    private _currentTarget: EventTarget;
+
+    private samples: InputSampleSequence = [];
+
+    constructor(identifier: number,
+                parsedObj: JSONInputSequence);
+    constructor(identifier: number,
+                initialTarget: EventTarget,
+                isFromTouch: boolean);
+    constructor(identifier: number,
+                obj: EventTarget | JSONInputSequence,
+                isFromTouch?: boolean) {
+      this.identifier = identifier;
+      if(obj instanceof EventTarget) {
+        this._currentTarget = obj;
+        this.isFromTouch = isFromTouch;
```

Comment:
The `isFromTouch` property is not marked as optional but the parameter is?

Reply:
TypeScript multi-signature constructors are... complicated to parse sometimes.  Note that the first version of the signature doesn't include the third parameter.  The first two signatures are the ones accessible for external code, while the final one is the constructor's "internal" signature, if that makes sense.  True constructor overloading is impossible in JS/TS, so we have to make do this way.

When the first form of the constructor is intended, this constructor will reuse the value found at `parsedObj.isFromTouch` - `parsedObj` will be a JSON parse of a previously-stringified instance.